doc/design/reader/low_level_reader.md

```diff
@@ -0,0 +1,71 @@
+# Paddle Low Level Data Reader
+
+This design doc will describe a way to implement C++ written data reader(feeder) to
+feed data to paddle when train or inference inorder to improve performance, and make
+best usage of CPU and GPU usage.
+
+This design can be used for both V2 API and new operator design.
+
+## Background
+
+We have sevaral issues talking about training with v2 API get bad training performance,
+like:
+- https://github.com/PaddlePaddle/Paddle/issues/3675 and
+- https://github.com/PaddlePaddle/Paddle/issues/4156 and
+- https://github.com/PaddlePaddle/Paddle/issues/3600.
+
+The current implementations for data reader can greatly simplify user defining how
+to parse and feed the data to neural network, but we need to copy data from python
+to C++ when feeding the data, and the data feeder will wait until the calculations
+of one mini-batch is done.
+
+In V1 implementations there are some considerations to improve data feeding
+performance, we can port these features to improve the reader.
+
+## Implementaion
+
+A low-level reader contains below components:
+
+1. Data parser:
+    A data parser will use `dlopen` to open user-defined parser plugin, and call
+    `parse` to convert a line of string to internal `Matrix` or `Tensor` object.
+1. User parser plugin:
+    User must write a C++ plugin to parse data to `Matrix` or `Tensor` object. The
+    plugin should implement one interface:
+    ```c++
+    // return Matrix* for v2 API call
+    Tensor* parse(const std::string& line);
+    ```
+1. Buffer:
+    A `DoubleBuffer` which is able to async load data when caculations are running.
+    We can select to use a general "memory buffer" or "mmap" buffer.
+1. Thread pool:
```

Comment:
In my opinion, the thread pool should be managed by the scheduler, independent of the reader OP.
For current V2 reader, we can use [paddle.reader.buffered](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/reader/decorator.py#L162) decorator to prefetch the data in the background thread. Please see example [here](https://github.com/PaddlePaddle/Paddle/blob/63bc2ff876724016062223bc3c966a4db43ae9c5/doc/design/reader/README.md#prefetch-data).